subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -401,11 +401,23 @@ private void configureCompatibilityOptions(DefaultJavaCompileSpec spec) {
             if (compileOptions.getRelease().isPresent()) {
                 spec.setRelease(compileOptions.getRelease().get());
             } else {
+                boolean didSetSourceOrTarget = false;
                 if (super.getSourceCompatibility() != null) {
                     spec.setSourceCompatibility(getSourceCompatibility());
+                    didSetSourceOrTarget = true;
                 }
                 if (super.getTargetCompatibility() != null) {
                     spec.setTargetCompatibility(getTargetCompatibility());
+                    didSetSourceOrTarget = true;
+                }
+                if (!didSetSourceOrTarget) {
+                    if (toolchain.getLanguageVersion().canCompileOrRun(9)) {
+                        spec.setRelease(toolchain.getLanguageVersion().asInt());
+                    } else {
+                        String version = toolchain.getLanguageVersion().toString();
```

Comment:
Consider extracting `toolchain.getLanguageVersion` as local variable to make it clear that these 3 all refer to the same value

Reply:
Another thought: would it make sense to use parts of this as the convention for the release property? So these properties are derived from the toolchain and not enforced by the toolchain presence? (I see what we might not want to consider that on `release`)

Reply:
Indeed, we could look into doing that ... later 😉 
But I am not sure it will work. We have so many conventions in that space already.